[Wei Xin] iP

Wei-Xinn opened the pull request:
# Duke

<blockquote>
 "Free your mind with Duke, your own friendly task manager!" - Anonymous
</blockquote>

Duke is your personal task manger that is:
- Text-based
- Easy to use
- **FREE!**

To use it, all you need to do is:
1.  Download Duke from [here](https://github.com/Wei-Xinn/ip/releases/tag/A-Jar)
2. Open Command Prompt
3. Enter `java -jar <Path of Duke>.jar`
4. **TADA**! Let it manage your tasks! 👍 👍 👍 

Features:
- [x] Managing Tasks
- [ ] Managing Deadlines (coming soon)
- [ ] Reminders (coming soon)

A good platform for you to practice Java too! Here is a snippet of it:
````
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
````

bamboolord committed:
Added Printer Class

Printer Class for output of Duke

bamboolord committed:
Fixed Printer Class Name

bamboolord committed:
Printer.print uses varargs

bamboolord committed:
Greet Echo Exit functionality

bamboolord committed:
Refactored Printer Class

bamboolord committed:
Refactored Input Parser

bamboolord committed:
Add Functionality

bamboolord committed:
List Functionality

bamboolord committed:
Added Task Class, InputerParser to handle multiple argumments

bamboolord committed:
Added Task Class, InputParser to handle multiple arguments

bamboolord committed:
arr changed to contain Task

bamboolord committed:
arr changed to contain Task

bamboolord committed:
Mark Functionality

bamboolord committed:
Mark Functionality

bamboolord committed:
unmark functionality

bamboolord committed:
unmark functionality

bamboolord committed:
Added TodoTask, EventTask, DeadlineTask

bamboolord committed:
Added TodoTask, EventTask, DeadlineTask

bamboolord committed:
TodoTask Creation

bamboolord committed:
TodoTask Creation

bamboolord committed:
Added Class Variable for Event and Deadline Task

bamboolord committed:
Added Class Variable for Event and Deadline Task

bamboolord committed:
DeadlineTask Creation and updated DeadlineTask toString method

bamboolord committed:
DeadlineTask Creation and updated DeadlineTask toString method

bamboolord committed:
EventTas Creation and updated EventTask toString method

bamboolord committed:
EventTas Creation and updated EventTask toString method

bamboolord committed:
Created DukeException Class

bamboolord committed:
Created DukeException Class

bamboolord committed:
Printer handle Empty List

bamboolord committed:
Printer handle Empty List

bamboolord committed:
Error Handling for Mark and Unmark Functions

bamboolord committed:
Error Handling for Mark and Unmark Functions

bamboolord committed:
Error Handling for Task Creation

bamboolord committed:
Error Handling for Task Creation

bamboolord committed:
Delete Functionality

bamboolord committed:
Delete Functionality

bamboolord committed:
Merge branch 'master' of github.com:Wei-Xinn/ip

bamboolord committed:
I/O Redirection Testing

bamboolord committed:
Testing now include Level-5 test cases

bamboolord committed:
Updated Testing to include Level-6 test cases

bamboolord committed:
Add Storage Class - File Save

bamboolord committed:
Revert "Add Storage Class - File Save"

This reverts commit 05522bc240d90d8e847e2e21a6660f97a74aff85.

bamboolord committed:
Add Storage Class - Save File Functionality

bamboolord committed:
Change save file path

bamboolord committed:
Fix save file to include delimiter

bamboolord committed:
Change Save File Format

bamboolord committed:
Add Load File Functionality

bamboolord committed:
Update Event and Deadline Tasks to use date

bamboolord committed:
Update Input Parser to process datetime

date is of format dd-M-yyyy hh:mm

bamboolord committed:
Add List Tasks on Date Functionality, Fix Deadline and EventTask

bamboolord committed:
Refactor InputParser

bamboolord committed:
Add List Tasks on date

bamboolord committed:
Add List before and after date

bamboolord committed:
Fix storage to be OS-independent

bamboolord committed:
Merge branch 'branch-Level-7'

bamboolord committed:
Merge branch 'branch-level-8'

bamboolord committed:
Fix bug after merge

bamboolord committed:
Refactor InputerParser to Parser

bamboolord committed:
Refactor Printer to Ui

bamboolord committed:
Revert "Refactor Printer to Ui"

This reverts commit b44fec28dcc5be0524a84f64cadfba8faa3498fe.

bamboolord committed:
Revert "Refactor InputerParser to Parser"

This reverts commit 97e51e5d6c2659d8a64f0eca00a6431df35bed74.

bamboolord committed:
Refactor fix

bamboolord committed:
Merge remote-tracking branch 'origin/master'

bamboolord committed:
Rename Printer to Ui and InputParser to Parser

bamboolord committed:
Add TaskList

bamboolord committed:
Package

bamboolord committed:
Fix Storage loading bug

bamboolord committed:
Add Junit Tests and Bug fix

bamboolord committed:
Add manifest file

bamboolord committed:
Add JavaDoc

bamboolord committed:
Fix to follow Coding Convention

bamboolord committed:
Fix Coding Standard: Long line in EventTask

bamboolord committed:
Add Find Task using keyword functionality

bamboolord committed:
Revert "Add Find Task using keyword functionality"

This reverts commit a7ca2a88fd0492a0968eb62a8b80e05dcd0f5f4d.

bamboolord committed:
Add Find Task Functionality

bamboolord committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/command/Parser.java

bamboolord committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/Parser.java

bamboolord committed:
Fix bug after merge

bamboolord committed:
Update JavaDoc in Parser

bamboolord committed:
Fix save tasks be triggered on changes

bamboolord committed:
Refactor Ui to print Duke logo

bamboolord committed:
Fix Variable Names in Duke.java

bamboolord committed:
Remove redundant statement in Duke.java

bamboolord committed:
Remove Empty Line

bamboolord committed:
Update class JavaDoc for Parser

bamboolord committed:
Update method JavaDoc in Ui

bamboolord committed:
Move list task based on date logic ot TaskList and optimization

bamboolord committed:
Add JavaDoc for getTasksBasedOnDate() method in TaskList.java

bamboolord committed:
Fix minor JavaDoc error

bamboolord committed:
Remove unecessary import

bamboolord committed:
Refactor TaskList add method to handle event creation

TaskList deals with the creation of event, so that other classes would not have to deal with the creation of events.
This enforces the Single Responsilibty Principle of OOP.

bamboolord committed:
Update JUnit Test due to update to TaskList.add

bamboolord committed:
Add JavaDoc for TaskList.addTask

bamboolord committed:
Update JavaDoc in Storage.java

bamboolord committed:
Remove unnecessary import

bamboolord committed:
Fix JavaDoc

Periods added and minor puncutation corrections

bamboolord committed:
Move Duke.java into package

bamboolord committed:
Add JavaFX

bamboolord committed:
Revert "Add JavaFX"

This reverts commit 0828a0b688323953c9b259436888ab62ae6352af.

bamboolord committed:
Add Gradle Support

bamboolord committed:
Add Gradle JUnit testing

bamboolord committed:
Fix CheckStyle for Duke Main

bamboolord committed:
Fix checkstyle for Duke Test

bamboolord committed:
Update build.gradle dependencies to include JavaFX

bamboolord committed:
Add Launcher Class

bamboolord committed:
Update Duke incorporate GUI

Update Duke incorporate GUI

The previous iteration of Duke is entirely text based, where the user input and Duke output are solely done in the terminal.
DialogBox.java is also added to facilitate the dialog boxes of the GUI.

2 classes, Parser and Ui are updated to incorporate this change.

Parser: User input will be captured from the GUI.

Ui: All output from Duke will be shown in the GUI.

With these changes, Duke can perform the way it was intended to, but with the addition of  a GUI.

bamboolord committed:
Fix File Saving and Loading

sharpstorm commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,190 @@
+                    String.format("There are now %d task(s) in the list.", taskList.getSize()));
+            return 1;
+        case "deadline":
+            taskList.addTask(inputString.substring(9).split("/by")[0].trim(), false,
+                    LocalDate.parse(inputString.split("/by")[1].substring(1), formatter), 1);
```
While I feel this is a very impressive one-liner, consider breaking it into multiple lines with intermediate variables for better readability.
```suggestion
            String taskDescription = inputString.substring(9).split("/by")[0].trim();
            LocalDate taskDate = LocalDate.parse(inputString.split("/by")[1].substring(1), formatter);
            taskList.addTask(taskDescription, false, taskDate, 1);
```

sharpstorm commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,190 @@
+                    String.format("There are now %d task(s) in the list.", taskList.getSize()));
+            return 1;
+        case "event":
+            taskList.addTask(inputString.substring(6).split("/at")[0].trim(), false,
+                    LocalDate.parse(inputString.split("/at")[1].substring(1), formatter), 2);
```
Consider breaking into multiple lines with intermediate variables for readability.
```suggestion
            String taskDescription = inputString.substring(6).split("/at")[0].trim();
            LocalDate taskDate = LocalDate.parse(inputString.split("/at")[1].substring(1), formatter);
            taskList.addTask(taskDescription, false, taskDate, 2);
```

sharpstorm commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,190 @@
+     * @param taskList The list of task to be checked with.
+     */
+
+    public void validate(String inputString, String action, String[] args, TaskList taskList) throws DukeException {
```
I like how you encapsulate the validation logic into a single function. Consider however, making it more clear what exactly you are validating.
```suggestion
    public void validateArguments(String inputString, String action, String[] args, TaskList taskList) throws DukeException {
```

sharpstorm commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,190 @@
+            break;
+        case "deadline":
+        case "event":
+            String flag = action.equals("deadline") ? "/by" : "/at";
```
I feel that `flag` isnt the right name here, consider renaming `flag` to be more descriptive.
```suggestion
            String delimiter = action.equals("deadline") ? "/by" : "/at";
```

sharpstorm commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,108 @@
+    public static int loadFile(String filePath, TaskList tl) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/M/yyyy");
+        try {
+            FileInputStream fs = new FileInputStream(filePath);
+            Scanner sc = new Scanner(fs);
+            StringBuilder ss = new StringBuilder();
```
Should a more intuitive variable name be used? `ss` does not really imply StringBuilder, perhaps `sb` would be more intuitive if you wanted to keep the abbrevation.
```suggestion
            StringBuilder sb = new StringBuilder();
```

sharpstorm replied:
Consider using longer names because the 2 alphabet variables make it difficult to remember which is which when you have so many of them, like differentiating `ss` from `sc`.
```suggestion
            FileInputStream fileInput = new FileInputStream(filePath);
            Scanner scanner = new Scanner(fs);
            StringBuilder stringBuilder = new StringBuilder();
```

sharpstorm commented on src/main/java/duke/task/DeadlineTask.java:
```diff
@@ -0,0 +1,33 @@
+ * A task that should be completed before the specified date.
+ */
+public class DeadlineTask extends Task {
+    private LocalDate d;
```
Consider using a more descriptive name for this attribute.
```suggestion
    private LocalDate date;
```

sharpstorm commented on src/main/java/duke/task/EventTask.java:
```diff
@@ -0,0 +1,34 @@
+ * A task that is happening on the specified date.
+ */
+public class EventTask extends Task {
+    private LocalDate d;
```
Consider using a more descriptive name for this attribute.
```suggestion
    private LocalDate date;
```

sharpstorm commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,190 @@
+            }
+            if (Arrays.stream(args).noneMatch(flag::equals)) {
+                throw new DukeException(String
+                        .format("%s flag not detected. Please specify date using %s!", flag, flag));
```
Consider changing how the line break is inserted to make `String.format` a unit for readability.
```suggestion
                throw new DukeException(String.format("%s flag not detected. "
                        + "Please specify date using %s!", flag, flag));
```

sharpstorm commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,108 @@
+     * @param folderName folder name to create the save file.
+     * @param fileName name of save file.
+     * @param arr list of tasks to be saved.
+     * @return 0: save success, -1: error encountered.
```
Should the first character describing parameters be capitalized?
```suggestion
     * @param folderName Folder name to create the save file.
     * @param fileName Name of save file.
     * @param arr List of tasks to be saved.
     * @return 0: save success, -1: error encountered.
```

sharpstorm commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,108 @@
+        try {
+            Files.createDirectories(Paths.get(folderName));
+            String filePath = folderName + "/" + fileName;
+            File myObj = new File(filePath);
```
Consider changing `myObj` to a more descriptive variable name.
```suggestion
            File savedFile = new File(filePath);
```

sharpstorm commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,126 @@
+     * @return A list of tasks that matches the specified date and comparisonType.
+     */
+    public ArrayList<Task> getTasksBasedOnDate(LocalDate date, int comparisonType) {
+        ArrayList<Predicate<Task>> taskTypePredicateList = new ArrayList<>();
+        ArrayList<Predicate<Task>> taskDatePredicateList = new ArrayList<>();
```
I like the naming convention, makes the purpose of the lists clear.

sharpstorm commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,91 @@
+     * ==========================================
+     * @param arr list of task to be displayed.
+     */
+    public void print(ArrayList<Task> arr, String header) {
```
Consider naming this `printWithHeader` instead to differentiate it from the normal `print` methods
```suggestion
    public void printWithHeader(ArrayList<Task> arr, String header) {
```

sharpstorm commented on src/test/java/DukeParserTest.java:
```diff
@@ -0,0 +1,122 @@
+    public void createTaskCommand() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/M/yyyy");
+        Parser parser = new Parser();
+        Ui p = new Ui();
```
Even as a temporary variable, i feel that `p` has no link with the `Ui` class. Consider changing it to `ui`. This suggestion applies to other test cases that follow a similar naming.
```suggestion
        Ui ui = new Ui();
```

sharpstorm commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,108 @@
+     * @param arr list of tasks to be saved.
+     * @return 0: save success, -1: error encountered.
+     */
+    public static int saveFile(String folderName, String fileName, ArrayList<Task> arr) {
```
Consider making it more explicit what the array is (since usage isnt immediate).
```suggestion
    public static int saveFile(String folderName, String fileName, ArrayList<Task> taskArr) {
```

sharpstorm commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,126 @@
+     * @param comparisonType The type of comparison to be performed.
+     * @return A list of tasks that matches the specified date and comparisonType.
+     */
+    public ArrayList<Task> getTasksBasedOnDate(LocalDate date, int comparisonType) {
```
Does `getTasksByDate` convey the same meaning more concisely?
```suggestion
    public ArrayList<Task> getTasksByDate(LocalDate date, int comparisonType) {
```

bamboolord committed:
Add JavaFx Fxml Views

bamboolord committed:
Fix minor bug in Duke.java

bamboolord committed:
Remove duke.txt

bamboolord committed:
Add Speech Bubble to GUI

cindyangXQ commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,108 @@
+                                    t.isDone() ? "T" : "F", t.getTaskName(), (
+                                            (EventTask) t).getDate().format(formatter)))
+                            .append(";");
+                }
```
You could consider writing a method for each kind of the tasks such as `String inFileRepresentation()`.

cindyangXQ commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,108 @@
+     * @param tl TaskList to be updated.
+     * @return 0: load success, -1: error encountered.
+     */
+    public static int loadFile(String filePath, TaskList tl) {
```
You might consider using `taskList` here instead of `tl`. It could be mistaken for`t1`, which is not preferable.

cindyangXQ commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,108 @@
+     *
+     * @param filePath path of the file to be read.
+     * @param tl TaskList to be updated.
+     * @return 0: load success, -1: error encountered.
```
I suppose this kind of expression using the colon should not be written in JavaDocs. I suggest `0 if loaded successfully and -1 if encountered errors.`

cindyangXQ reviewed (commented):
Sorry for the late comment. The code is well written. Logic clearly expressed.

bamboolord committed:
Fix minor issues

bamboolord committed:
Fix Checkstyle

bamboolord committed:
Add assertions at critical logic areas

bamboolord committed:
Refactor Task classes to support TaskList toFileString method

Storage Class requires all the classes under duke.Task package to be imported.

This is not ideal as it in addition to the places unecessary dependancy on the individual tasks classes, violating the Single Responsbility Principle of OOP.
It woud hence be more ideal for the storage class to only interact with the TaskList.

By providing a toFileString method, which coverts the tasks to a specificed file string format, TaskList can be improved to include a toFileString method
which converts its list of Tasks to a specified String format for file saving.

This ensures that the Storage class has minimal dependancy on the Task Classes, and only interacts with the TaskList class.
This potentially makes future debugging easier due to the reduced depedancies of classses.

bamboolord committed:
Add line breaks to improve readability

bamboolord committed:
Rename variable names in EventTask and DeadlineTask

bamboolord committed:
Remove unecessary line break

bamboolord committed:
Refactor UI class to be static

Methods under the UI class are not static, and would therefore require a instance to be created before its method could be accessed.

After reviewing the functionalities of the program, there is no need for there to be multiple instances of Ui.

By changing the methods of Ui to be static emulates the implementation of a singel Ui instance throughout the classes, so that any call refers to the same instance.

This removes the hassle of needing to create a Ui class just to acess its methods.

bamboolord committed:
Refactor Parser methods to be static

Similar to the UI class previously, methods under the Parser class are not static, and would therefore require a instance to be created before its method could be accessed.

The role of the Parser class is to simply provide the run and validate method, which does not gain any functionality from a instance being created.
Even when considering any future possible changes, there should not be a need where multiple Parser instances will be needed.

By changing the methods of Parser, namely validate and run to be static, it emulates the implementation of a singel Parser instance throughout the classes, ensuring consistency.

This removes the hassle of needing to create a Parser class just to acess its methods.

bamboolord committed:
Change Parser.validate to be private

Wei-Xinn committed:
Merge pull request #2 from Wei-Xinn/CodeQuality

Code quality

Wei-Xinn committed:
Merge branch 'master' into Assertions

Wei-Xinn committed:
Merge pull request #3 from Wei-Xinn/Assertions

Add assertions at critical logic areas

bamboolord committed:
Fix minor JavaDoc format

bamboolord committed:
Fix variable name

bamboolord committed:
Fix checkstyle

bamboolord committed:
Set up Continuous Integration

bamboolord committed:
Add check functionality for duplicate task name

bamboolord committed:
Improve code readability

bamboolord committed:
Fix variable names

bamboolord committed:
Fix variable name

bamboolord committed:
Remove unused Task constructors

bamboolord committed:
Add JavaDoc

bamboolord committed:
Add scalable dialogbox

bamboolord committed:
Fix build.gradle

bamboolord committed:
Remove redundant new line

bamboolord committed:
Fix checkstyle

bamboolord committed:
Change Window title form Duke to Jerry

bamboolord committed:
Clean up code

bamboolord committed:
Fix DukeParserTest

bamboolord committed:
Add right margin for user image

bamboolord committed:
Fix toString format

bamboolord committed:
Update Readme

bamboolord committed:
Fix format for Readme

bamboolord committed:
Update Readme

bamboolord committed:
Fix minor Readme errors

bamboolord committed:
Fix Readme

bamboolord committed:
Fix fxml files

bamboolord committed:
Fix application to close on bye
